File: web/src/lib/crypto/decrypt.ts
import nacl from 'tweetnacl';
import { encodeUTF8, decodeBase64 } from 'tweetnacl-util';
import type { EncryptedPayload } from '@/types/chat';

/**
 * 메시지 복호화 (nacl.box.open - XSalsa20-Poly1305)
 * 사전 계산된 공유 비밀 사용
 */
export function decryptMessage(
  payload: EncryptedPayload,
  sharedSecret: Uint8Array
): string | null {
  const ciphertext = decodeBase64(payload.ciphertext);
  const nonce = decodeBase64(payload.nonce);
  const decrypted = nacl.box.open.after(ciphertext, nonce, sharedSecret);

  if (!decrypted) {
    return null;
  }

  return encodeUTF8(decrypted);
}

File: mobile/lib/core/crypto/decrypt.dart
import 'dart:convert';

import 'package:pinenacl/x25519.dart';

import 'encrypt.dart';

/// 메시지 복호화 (web: nacl.box.open.after)
/// 변조 감지 시 null 반환
String? decryptMessage(EncryptedPayload payload, Uint8List sharedSecret) {
  try {
    final ciphertext = base64Decode(payload.ciphertext);
    final nonce = base64Decode(payload.nonce);

    final box = SecretBox(sharedSecret);
    final encrypted = EncryptedMessage(
      nonce: Uint8List.fromList(nonce),
      cipherText: Uint8List.fromList(ciphertext),
    );
    final plaintext = box.decrypt(encrypted);

    return utf8.decode(plaintext);
  } catch (_) {
    return null;
  }
}
